### DISCUSSION TITLE
Question regarding multiple render passes

### DISCUSSION THREAD

**Discussion by ChristianDumancic** [skaarj1989/FrameGraph#7] [Q&A] [answered]
First of all, excellent library! It helped tremendously to simplify my rendering code. I do however have a question about best practice in certain case.

In `FrameGraph::addCallbackPass()` function, second parameter is function called when we are setting up the framegraph nodes in order to designate which resources are being created, read or written to.

So this:

`data.something = builder.read(someOtherPass.something);`

will mark input resource as readable - one used to read from, this:

```
data.something = builder.create<FrameGraphTexture>("Something", desc);
data.something = builder.write(data.something);
```

creates output resource and marks it as writeable, increments version by 1. How do you manage resource which will be written to in multiple passes?

As a simple example let's say we wish to render scene using forward renderer -

- first pass would render opaque meshes,
- second pass transparent meshes,
- third pass particles,
- fourth pass GUI.

In this case if we wish to render multiple times onto the exact same surface, if we use this:

`data.something = builder.write(data.something);`

framegraph will increment version by 1 each time we call `builder.write()` so it will create multiple render targets thus rendering opaque meshes onto RT1, transparent meshes onto RT2, particles on RT1 again and then GUI onto RT2.

If we do not use builder.write() but simply try and use the same render target pass will be culled, if for example particles pass doesn't have any inputs and is not marked with `builder.write()` as to have any outputs. Setting `builder.setSideEffect()` will trigger asserts about resource usage.

Is there any way to solve this? I'm probably missing something simple but can figure out what. Could be I'm just not using it correctly...

**Comment by skaarj1989** [comment, marked as answer]
> framegraph will increment version by 1 each time we call builder.write() so it will create multiple render targets thus rendering opaque meshes onto RT1, transparent meshes onto RT2, particles on RT1 again and then GUI onto RT2.

No. Whenever you call `builder.write()` you get a new `FrameGraphResource` handle with incremented version, but the underlying resource is the same as an input resource.
Bumping the version enforces correct execution order.

> How do you manage resource which will be written to in multiple passes?

Look at [here](https://github.com/skaarj1989/FrameGraph-Example/blob/5861876f390205f2d17999c9b1e1c6efe70a1f18/src/WorldRenderer.cpp#L227C9-L227C9).
And [here](https://github.com/skaarj1989/FrameGraph-Example/blob/5861876f390205f2d17999c9b1e1c6efe70a1f18/src/Passes/SkyboxPass.cpp) you can find a simple pass that draws a skybox over a render target (after opaques deferred pass).

**Reply by ChristianDumancic** [reply]
Works now, it was simple mistake - in one place I was mixing framegraph calls with old hardcoded ones and one render target was set wrongly. Thanks for the help. :)
